### ISSUE TITLE
imr: switch to run-time (open-coded) description of structures

### ISSUE THREAD

**Issue by denesb** [scylladb/scylla#5578]
When the IMR infrastructure was originally proposed there was a long discussion on whether the description of structures (e.g. cells) should be at least partially compile-time (via template meta-programming) or whether it should be entirely runtime (open-coded) [[1]](https://groups.google.com/d/topic/scylladb-dev/wsGtRT3pL-E/discussion).

The arguments for-against the compile-time description were:

Pros:
* Human readable definition of the building-blocks, available up-ahead and statically.
* More efficient code.
* Easier to change anything as the compiler generates the code.
* Easier way forward to LLVM JIT compiled (de)serialization.

Cons:
* Complex, very hard to understand and modify implementation (template meta-programming).
* Low cost/benefit ratio, as most things are known only at runtime.
* Poor diagnostics.
* Slow compile times.

After some back-and-forth we were convinced that the the pros outweigh the cons. Indeed compile times didn't prove to be an issue and most building blocks of cells could be nicely described compile-time with `imr::variant` providing the necessary runtime switching between the possible options. We were also told that tools will likely catch up soon with templates. The sad reality is that they still haven't. Also the complexity and unreadability of the implementation remained a serious issue, exacerbated with the original author no longer being with us.
After some mileage with the IMR code additional problems surfaced:
* Debugability is horrible as `gdb` fails to provide accurate source code mapping inside the `imr` code, one is left to either debug ASM or try to guess where the problem is based on code addresses of the frames.
* Although the definitions of new building blocks can be done easily, things quickly become very tricky as it is very easy to write context objects that only handle parts of the fields, producing corrupt data. Generally it is easy to make mistakes as template meta-programming is not type-safe. This was not yet seen in the wild (AFAIC), I only bumped into this when experimenting with moving containers to IMR.

IMR was supposed to become the infrastructure for all our building blocks, from cells all the way to rows, keys and maybe even partitions. However development has halted as developers are reluctant to touch the code and maintainers are even more reluctant to make even more of our building blocks this complex.

To sum it up, after reconsidering the question, with some mileage under our belt, it seems that runtime description of types would be a better option for such mundane reasons as readability, maintainability and debugability. There are no surprises here, we've known/suspected all the pros/cons from the beginning, it is just their cost/benefit ration that we miscalculated.

**Comment by kbr-** [comment]
> Complex, very hard to understand and modify implementation (template meta-programming)

Does the mere fact that it uses template meta-programming make it complex?
Non-template code can be easy or hard to understand, depending on how it's written; the same stands for template code.

C++ has evolved a lot since IMR was first implemented, introducing constructs like concepts or constexpr if, which perhaps should make some of these things easier.

Maybe IMR simply requires a bit of refactoring, not a complete rewrite to runtime descriptions? I'd actually be interested in digging into this a bit.

**Comment by avikivity** [comment]
It's not just the language dialect (and it does use concepts).

IMO it is a bad tradeoff for a database. It introduces a lot of complexity, but can see only a limited use since the schema is known at runtime only.

I'm dying to have http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1609r3.html so we can write


```c++
schema& s = ...;
auto make_row_for_this_schema = dynamic_function_template_instantiation<make_row>(s);
```

and get a make_row that stores five ints if that's the schema, or eight blobs, or whatever.

**Comment by denesb** [comment]
> > Complex, very hard to understand and modify implementation (template meta-programming)
> 
> Does the mere fact that it uses template meta-programming make it complex?

The sheer amount of it. Just have a look at stuff in the `imr/` and `data/` folders and judge it for yourself.

> Non-template code can be easy or hard to understand, depending on how it's written; the same stands for template code.

True to some degree, but the playing fields is not level. Template meta-programming has a much more restricted tool-set available to it, making it far less expressive. The compiler is also if little help with diagnostics. Template meta-programming also shares the macros' bad traits of not being able to see the generated code.

> 
> C++ has evolved a lot since IMR was first implemented, introducing constructs like concepts or constexpr if, which perhaps should make some of these things easier.

All these were available when IMR was implemented and they are used too.

> 
> Maybe IMR simply requires a bit of refactoring, not a complete rewrite to runtime descriptions? I'd actually be interested in digging into this a bit.

Have a look at the folders I mentioned above.

**Comment by denesb** [comment]
I'm just throwing this out here, an IMR frame in gdb:
```
#11 data::cell::make_collection<fragment_range_view<bytes_ostream>, void>(fragment_range_view<bytes_ostream>)::{lambda(auto:1&&, auto:2&&)#1}::operator()<imr::internal::structure_sizer<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation>, {lambda(auto:1&&, auto:2&&)#1}::operator()::member<data::cell::tags::flags, {lambda(auto:1&&, auto:2&&)#1}::operator()::flags<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::collection, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::counter_update, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::empty, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::external_data> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell, {lambda(auto:1&&, auto:2&&)#1}::operator()::variant<{lambda(auto:1&&, auto:2&&)#1}::operator()::member<data::cell::tags::flags, {lambda(auto:1&&, auto:2&&)#1}::operator()::flags<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::collection, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::counter_update, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::empty, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::external_data> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell, {lambda(auto:1&&, auto:2&&)#1}::operator()::structure<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod<long> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::optional<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::ttl, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiry, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::optional<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::ttl, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiry, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::dead, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod>, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod>, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value, {lambda(auto:1&&, auto:2&&)#1}::operator()::buffer<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::variable_value, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > > > > > > > > > > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::flags, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > > > > > > > > > >, {lambda(auto:1&&, auto:2&&)#1}::operator()::alloc::object_allocator::sizer&>(imr::internal::structure_sizer<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation>, {lambda(auto:1&&, auto:2&&)#1}::operator()::member<data::cell::tags::flags, {lambda(auto:1&&, auto:2&&)#1}::operator()::flags<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::collection, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::counter_update, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::empty, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::external_data> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell, {lambda(auto:1&&, auto:2&&)#1}::operator()::variant<{lambda(auto:1&&, auto:2&&)#1}::operator()::member<data::cell::tags::flags, {lambda(auto:1&&, auto:2&&)#1}::operator()::flags<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::collection, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::counter_update, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::empty, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::external_data> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell, {lambda(auto:1&&, auto:2&&)#1}::operator()::structure<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod<long> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::optional<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::ttl, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiry, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::optional<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::live, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::ttl, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiry, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::dead, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod>, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod>, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value, {lambda(auto:1&&, auto:2&&)#1}::operator()::buffer<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::expiring, {lambda(auto:1&&, auto:2&&)#1}::operator()::pod> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::variable_value, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::atomic_cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > > > > > > > > > > > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::flags, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::cell<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_size, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned int> >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > >, imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::data, {lambda--Type <RET> for more, q to quit, c to continue without paging--
(auto:1&&, auto:2&&)#1}::operator()::member::fixed_value<imr::structure_sizer_continuation<{lambda(auto:1&&, auto:2&&)#1}::operator()::no_op_continuation><{lambda(auto:1&&, auto:2&&)#1}::operator()::member::pointer, {lambda(auto:1&&, auto:2&&)#1}::operator()::tagged_type<{lambda(auto:1&&, auto:2&&)#1}::operator()::member::value_data, {lambda(auto:1&&, auto:2&&)#1}::operator()::member::timestamp<unsigned char*> > > > > > > > > > >&&, {lambda(auto:1&&, auto:2&&)#1}::operator()::alloc::object_allocator::sizer&) const (allocations=..., allocations=..., serializer=..., this=0x7ffffc3b4bb8) at ./data/cell.hh:305
305                     .done();
```

**Comment by denesb** [comment]
@kbr- there is a high-level documentation on the design in https://github.com/scylladb/scylla/blob/master/docs/in_memory_representation.md
